man/mlterm.1
.\" mlterm.1   -*- nroff -*-
.TH MLTERM 1 "2026-07-05"
.SH NAME
mlterm \- Multi Lingual TERMinal emulator on X
.SH SYNOPSIS
.B mlterm
.RB [\fIoptions\fP]
.\" ********************************************************************
.SH DESCRIPTION
\fBmlterm\fP is a multi-lingual terminal emulator written from
scratch, which supports various character sets and encodings
in the world and complex characters such as double width for East
Asian, combining for Thai, Vietnamese, and so on, and bi-direction for
Arabic and Hebrew.  Indic scripts are experimentally supported.
It also supports various unique feature such
as anti-alias using FreeType, multiple XIM, multiple windows,
scrollbar API, scroll by mouse wheel, automatic selection of
encoding, daemon mode, and so on.
.PP
Supported encodings are:
ISO-8859-[1-11], ISO-8859-[13-16], TIS-620 (same as ISO-8859-11), KOI8-R,
KOI8-U, KOI8-T, GEORGIAN-PS, TCVN5712, ISCII_(ASSAMESE|BENGALI|GUJARATI|
HINDI|KANNADA|MALAYALAM|ORIYA|PUNJABI|TAMIL|TELUGU), VISCII, CP125[0-8],
CP874, EUC-JP, EUC-JISX0213, Shift_JIS, Shift_JISX0213,
ISO-2022-JP[1-3], EUC-KR, UHC, JOHAB, ISO-2022-KR, GB2312 (EUC-CN), GBK,
GB18030, ISO-2022-CN, HZ, EUC-TW, BIG5, BIG5HKSCS, and UTF-8.
If you have already set locale (for example LANG variable;
see \fBlocale\fR(7) for detail) \fBmlterm\fR will automatically select
proper encoding.
.PP
.\" ********************************************************************
.SH OPTIONS
Note that \fIbool\fR is to be substituted by \fBtrue\fR or \fBfalse\fR.
.TP
\fB\-A\fR, \fB\-\-aa\fR(=\fIbool\fR)
Use anti-aliased fonts. This option works only with Xft or cairo for now.
The default is \fBfalse\fR.
.TP
\fB\-B\fR, \fB\-\-sbbg\fR=\fIcolor\fR
Specify a background color of a scrollbar.
A valid value for \fIcolor\fR is a color name or a RGB value.
The color name should be defined in rgb.txt or "\fBcolor\fR" configuration file.
The RGB value's format should be "#RRGGBB", "#RRGGBBAA", "rgb:RR/GG/BB" or "rgba:RR/GG/BB/AA".
.TP
\fB\-C\fR, \fB\-\-ctl\fR(=\fIbool\fR)
Enable complex text layouting on UTF8 encoding to support indic scripts and
RTL (right-to-left) languages such as Arabic and Hebrew.
The default is \fBtrue\fR.
(If mlterm -v outputs "no-indic" or "no-bidi", indic scripts or RTL languages
are not supported.)
.TP
\fB\-E\fR, \fB\-\-km\fR=\fIencoding\fR
Specify encoding.
Valid encodings are listed in \fBDESCRIPTION\fR section
above in this man page.
\fBAUTO\fR makes mlterm determine the according according to
the current locale (default \fBAUTO\fR).
.TP
\fB\-F\fR, \fB\-\-sbfg\fR=\fIcolor\fR
Specify a foreground color of a scrollbar.
See \fB\-\-sbbg\fR option for valid values.
.TP
\fB\-G\fR, \fB\-\-vertical\fR=\fImode\fR
Specify vertical writing mode.
\fBcjk\fR for RTL vertical writing and \fBmongol\fR for LTR one.
The default is \fBnone\fR which means horizontal writing mode.
.TP
\fB\-H\fR, \fB\-\-bright\fR=\fIpercent\fR
Brightness of background images in percent.
See \fB\-p\fR option for details of background images.
The default is \fB100\fR (keep original).
.TP
\fB\-I\fR, \fB\-\-icon\fR=\fIname\fR
Specify a name to be used when a mlterm window is iconified.
The default is "\fBmlterm\fR".
.TP
\fB\-J\fR, \fB\-\-dyncomb\fR(=\fIbool\fR)
Enable dynamic character combining.
"Dynamic" means that combining characters are stored in
without combining but they are displayed using combined form.
This affects calculation of column position,
i.e., a pair of base character and combining character
is counted to be two columns in this mode, while it is counted
to be one column in the normal mode.  Under this option,
a (logical) column number and a character one-to-one correspondence.
even for combining characters (though not for fullwidth characters;
see \fB\-Z\fR/\fB\-\-multicol\fR option for handling of fullwidth characters).
Thus, this mode enables you to use combining characters with
software which does not support combining characters.
The default is \fBfalse\fR.
.TP
\fB\-K\fR, \fB\-\-metakey\fR=\fIkey\fR
Specify a key to be interpreted as a META key.
Valid \fIvalue\fRs are: \fBalt\fR, \fBmeta\fR, \fBhyper\fR,
\fBsuper\fR, \fBmod1\fR, \fBmod2\fR, \fBmod3\fR, \fBmod4\fR,
and \fBnone\fR.   The default is \fBnone\fR.

See \fB\-k\fR option also.
.TP
\fB\-L\fR, \fB\-\-ls\fR(=\fIbool\fR)
Whether to use login shell or not.  The default is \fBfalse\fR.
.TP
\fB\-M\fR, \fB\-\-im\fR= \fIinput method\fR : \fB[\fR\fI arguments\fR \fB... ]\fR
Specify an input method.

.PP
.RS
.B Examples:
.TP
\-\-im=default
Use the default input method provided by the system.
.TP
\-\-im=xim
Use XIM with the default XIM server specified by standard way (i.e.,
\fBXMODIFIERS\fR environmental variable).
.TP
\-\-im=xim:Ami
Use XIM with Ami on the system locale.
.TP
\-\-im=xim:kinput2:ja_JP.EUC-JP
Use XIM with kinput2 on ja_JP.EUC-JP locale.
.TP
\-\-im=kbd:arabic
Use keyboard mapping input method in Arabic.
.TP
\-\-im=kbd:hide:arabic
Use keyboard mapping input method without showing the status screen.

"kbd" input method converts US keyboard characters to other keyboard layouts.
It is assumed to be used in environments with limited functionality,
such as older OSs that support few keyboard layouts.
Available values for "kbd" input method argument are "arabic", "hebrew",
"de", "fr", "es", "pt", "jp106", "isciiassamese", "isciibengali",
"isciigujarati", "isciihindi", "isciikannada", "isciimalayalam", "isciioriya",
"isciipunjabi", "isciitamil" and "isciitelugu".
.TP
\-\-im=uim
Use uim with the default conversion engine.
.TP
\-\-im=uim:prime
Use uim with prime conversion engine.
.TP
\-\-im=m17nlib:ru
Use m17n library in Russian.
.TP
\-\-im=m17nlib:or:itrans
Use m17n library in Oriya using ITRANS method.
.TP
\-\-im=scim
Use SCIM.
.TP
\-\-im=ibus
Use IBus with the default conversion engine.
.TP
\-\-im=ibus:anthy
Use IBus with anthy conversion engine.
.TP
\-\-im=fcitx
Use Fcitx.
.TP
\-\-im=canna
Use Canna.
.TP
\-\-im=wnn
Use Freewnn.
.TP
\-\-im=wnn:foo.bar
Use Freewnn with jserver at foo.bar host.
(JSERVER environmental variable is also available.)
.TP
\-\-im=skk
Use SKK.
.TP
\-\-im=skk:dict=foo.bar:utf8,sskey=\\x3b
Use SKK with the use of utf8 skk server at foo.bar host and semicolon key as sticky shift key.
(SKK_DICTIONARY and SKK_STICKY_SHIFT_KEY environmental variable are also available.)
.TP
\-\-im=iiimf
Use IIIMF in the system language.
.TP
\-\-im=iiimf:ar
Use IIIMF in Arabic.
.TP
\-\-im=iiimf:ja:CannaLE
Use IIIMF in Japanese using CannaLE language engine.
.TP
.TP
\-\-im=none
Don't use input method.
.RE
.TP
\fB\-N\fR, \fB\-\-name\fR=\fIname\fR
Specify application name. (WM_CLASS property)
The default is "\fBmlterm\fR".
.TP
\fB\-\-role\fR=\fIrole\fR
Specify application role. (WM_WINDOW_ROLE property)
The default is undefined.
.TP
\fB\-O\fR, \fB\-\-sbmod\fR=\fImode\fR
Specify the side to show a scrollbar.
\fBleft\fR for left side and \fBright\fR for right side.
\fBnone\fR turns off a scrollbar.
\fBautohide\fR shows a scrollbar only if mouse pointer is
at the right edge of the screen.
The default is \fBleft\fR.
.TP
\fB\-Q\fR, \fB\-\-vcur\fR(=\fIbool\fR)
Change interpretation of cursor keys to be natural
in vertical writing mode.
This means that up and down arrow keys are treated as
backward (left arrow in horizontal LTR) and forward
(right arrow in horizontal LTR), respectively.
In \fBcjk\fR \fB\-G\fR/\fB\-\-vertical\fR mode,
left and right arrow keys are also treated as next line
(down arrow in horizontal LTR) and previous line
(up arrow in horizontal LTR), respectively, while
vice versa in \fBmongol\fR mode.
The default is \fBtrue\fR.
.TP
\fB\-R\fR, \fB\-\-fsrange\fR=\fIrange\fR
Set acceptable range of font size.
The format is "\fIminsize\fR-\fImaxsize\fR", where \fIminsize\fR and
\fImaxsize\fR are font sizes in pixel (default \fB1-200\fR for mlterm-fb,
mlterm-wl or mlterm-sdl2 with freetype support, \fB1-10000\fR otherwise).
The GUI configurator and other means for setting fontsize should honor
the range.
.TP
\fB\-S\fR, \fB\-\-sbview\fR=\fIname\fR
Select a type of scrollbar.  See \fBSCROLLBAR\fR section below
for details.  The default is "\fBsimple\fR" which means the
built-in simple scrollbar.
.TP
\fB\-T\fR, \fB\-\-title\fR=\fIname\fR
Specify a title for a mlterm window.
The default is "\fBmlterm\fR".
.TP
\fB\-\-locktitle\fR(=\fIbool\fR)
Lock the title for a mlterm window.  If a title is locked then
OSC 0, 1 and 2 esc sequences to change it are ignored.
The default is \fBfalse\fR.
.TP
\fB\-U\fR, \fB\-\-viaucs\fR(=\fIbool\fR)
Force to convert a selection (i.e., copy-and-paste strings) whose
type is not UTF8_STRING to the current mlterm encoding via Unicode.
See \fBSELECTION\fR section below for detail.
The default is \fBfalse\fR.
.TP
\fB\-V\fR, \fB\-\-varwidth\fR(=\fIbool\fR)
Use variable column width.  You may want to use this
option when you use proportional fonts.
The default is \fBfalse\fR.
.TP
\fB\-W\fR, \fB\-\-sep\fR=\fIcharacterlist\fR
Delimiter characters used for word selection, which are
consulted when you double-clicked mlterm, to define what
is a word.
The default is "\fB ,.:;/|@()[]{}\fR")
.TP
\fB\-X\fR, \fB\-\-alpha\fR=\fInumber\fR
Alpha in pseudo or true transparent.
The default is \fB255\fR.
.TP
\fB\-Y\fR, \fB\-\-decsp\fR(=\fIbool\fR)
Use dynamically composed line drawing character set
of DEC special.
The default is \fBfalse\fR.
This overrides DEC_SPECIAL in "font" configuration file,
while DEC_SPECIAL in "aafont" (for Xft or cairo) is always
overridden.
.TP
\fB\-Z\fR, \fB\-\-multicol\fR(=\fIbool\fR)
Treat fullwidth characters (east Asian characters in most cases;
which occupies two columns on the screen) as they occupy two logical
columns.  It is the de-facto standard way to handle
fullwidth characters in east Asian terminal emulators
(XFree86 xterm and  kterm, cxterm, hanterm, rxvt, eterm)
and other systems such as MS\-DOS, PC\-9801, and so on.
In most fonts, the glyphs of fullwidth characters are designed
assuming that their width are twice of normal characters and
won't display correctly without this option.
The default is \fBtrue\fR.
.TP
\fB\-0\fR, \fB\-\-crbg\fR=\fIcolor\fR
Specify background color for cursor (default is same to foreground color).
Valid values for \fIcolor\fR are color names defined in rgb.txt and
color rgb string whose format is "#RRGGBB", "#RRGGBBAA",
"rgb:RR/GG/BB" or "rgba:RR/GG/BB/AA".
.TP
\fB\-1\fR, \fB\-\-wscr\fR=\fIpercent\fR
Specify actual window width, by percentage against
calculated value by multiplying font width by column number.
This is useful when you use a proportional font which includes
some glyphs with exceptionally large width, i.e., much
larger "maximum width" than your expectation.
In vertical mode this option changes actual window height.
The default is \fB100\fR.
.TP
\fB\-3\fR, \fB\-\-contrast\fR=\fIpercent\fR
Contrast of background image in percent.
See \fB\-p\fR option for details of background image.
The default is \fB100\fR.
.TP
\fB\-4\fR, \fB\-\-gamma\fR=\fIpercent\fR
Gamma of background image in percent.
See \fB\-p\fR option for details of background image.
The default is \fB100\fR.
.TP
\fB\-5\fR, \fB\-\-big5bug\fR(=\fIbool\fR)
Enable a workaround for Big5 CTEXT bugs (which had been existed until XFree86 4.1.0).
This affects Big5 selections (i.e., copy-and-paste strings) in
COMPOUND_TEXT format which \fBmlterm\fR sends.
The default is \fBfalse\fR.
.TP
\fB\-6\fR, \fB\-\-stbs\fR(=\fIbool\fR)
Don't exit backscroll mode when console applications output something.
The default is \fBfalse\fR.
.TP
\fB\-7\fR, \fB\-\-bel\fR=\fImode\fR
Behavior when BEL (0x07) is received. \fBsound\fR for beep
, \fBvisual\fR for blanking screen and \fBsound|visual\fR for the both.
The default is \fBnone\fR which ignores BEL.
.TP
\fB\-8\fR, \fB\-\-88591\fR(=\fIbool\fR)
Use ISO8859-1 fonts for US-ASCII part of various encodings.
.TP
\fB\-9\fR, \fB\-\-crfg\fR=\fIcolor\fR
Specify foreground color for cursor (default is same to background color).
Valid values for \fIcolor\fR are color names defined in rgb.txt and
color rgb string whose format is "#RRGGBB", "#RRGGBBAA",
"rgb:RR/GG/BB" or "rgba:RR/GG/BB/AA".
.TP
\fB\-$\fR, \fB\-\-mc\fR=\fInumber\fR
Doubleclick/tripleclick interval in millisecond.
The default is \fB250\fR.
.TP
\fB\-%\fR, \fB\-\-logseq\fR(=\fIbool\fR)
Enable logging. Contents of stream received by mlterm will be logged under

ChangeLog
2026-09-27  Araki Ken <arakiken@user.sf.net>

	* c_sixel.c:
	  - Limit 'rep' to MAX_REPEAT.
	  - Allocate 'pixels' according to 'pix_y + 6' and 'pix_x + rep'.
	  - Ensure that 'color' is <=128 in ormode.
	  - Clear the all allocated memory of 'ext_palette' by memset().

2026-09-26  Araki Ken <arakiken@user.sf.net>

	* ui_display_linux.c: get_event_device_num() searches multiple mouse
	  devices.

	* ui_font_ft.c: get_dpi() checks if XResourceManagerString() returns
	  NULL to avoid a segfault when starting mlterm in environments where X
	  server is not working.

	* fb/ui_display_linux.c: Fix conversion from EV_ABS to EV_REL.

	* ui_screen.c: Pass (event->state & screen->mod_ignore_mask) to
	  ui_shortcut_str() and ui_shortcut_match().

	* ui_shortcut.c: Remove NumLock (Mod2Mask) hack from ui_shortcut_str()
	  and ui_shortcut_match().

2026-09-21  Araki Ken <arakiken@user.sf.net>

	* ui_display_linux.c: Refresh the screen every time the cursor moves.

	* ui_screen.c: Call ui_stop_selecting() when pressing ESC or q in selecting
	  text in copymode.

2026-09-20  Araki Ken <arakiken@user.sf.net>

	* PROTOCOL, README.ja, man/mlterm.1, etc/main, etc/main.ja: Update.

	* ui_screen.h: Define USE_IM_CURSOR_COLOR by default.

	* ui_screen.c: Add im_cursor_color to OSC 5379 - 5381.

	* im_uim.c:
	  - Enable mode_update() by default.
	  - Implement switch_mode(), which calls uim_set_mode().

	* inputmethod/*/Makefile.in: Remove @IM_CFLAGS@.

2026-09-19  Araki Ken <arakiken@user.sf.net>

	* PROTOCOL, README.ja, man/mlterm.1, etc/main, etc/main.ja: Update.

	* ci.yml: Fix PKG_CONFIG_PATH in build-solaris.

2026-09-13  Araki Ken <arakiken@user.sf.net>

	* ui_emoji.c: Execute ui_emoji_set_path("/boot/system") in HaikuOS in TEST_ui_emoji().

	* ci.yml:
	  - Remove libXft in build-netbsd.
	  - Remove sudo in debian:11.

2026-09-13  Araki Ken <arakiken@user.sf.net>

	* ci.yml: deb.debian.org, security.debian.org -> archive.debian.org in
	  /etc/apt/sources.list in debian:11.

	* vte.c: blink_cursor -> cursor_style in vte_terminal_init().

	* mlconfig/main.c: #ifdef GTK_CHECK_VERSION -> #if GTK_CHECK_VERSION

	* ui_main_config.[ch]:
	  - Add --cursor / cursor_style option.
	  - blink_cursor -> cursor_style in ui_main_config_t.
	  - Rearrange ui_main_config_t.

	* vt_parser.c:
	  - Add get_cursor_style_name() and vt_get_cursor_style_by_name().
	  - ignored_csi_list option supports \d which matches number parameters in CSI.

	* main_loop.c: Add --decscusr/allow_decscusr option.

2026-09-06  Araki Ken <arakiken@user.sf.net>

	* ui_draw_str.c: Fix draw_drcs() again.
	  (https://github.com/arakiken/mlterm/issues/176#issuecomment-5463271145)

2026-09-05  Araki Ken <arakiken@user.sf.net>

	* ui_draw_str.c: Fix draw_drcs() not to draw garbage.
	  (https://github.com/arakiken/mlterm/issues/176#issuecomment-5463271145)

	* wayland/ui_display.c: keyboard_leave() sets wlserv->prev_key.type = 0.
	  (https://github.com/arakiken/mlterm/issues/177)

2026-08-31  Araki Ken <arakiken@user.sf.net>

	* ci.yml: Add test of building libvte compatible library to build-solaris.

	* mlterm-menu/Makefile.in, mlconfig/Makefile.in: Add LIBS2_beos.

	* zmodem.c: Fix buffer overflow caused by sprintf().

2026-08-29  Araki Ken <arakiken@user.sf.net>

	* beos.cpp: Define PTHREAD_RECURSIVE_MUTEX_INITIALIZER, B_HANGUL and
	  B_HANGUL_HANJA macros which are dropped in r1beta6.

	* ui_font_manager.c: ui_get_font() returns a font with size_attr for
	  DRCS charset.

2026-08-23  Araki Ken <arakiken@user.sf.net>

	* configure.in: Check libnetwork for @SOCK_LIBS@ if libsocket doesn't exist.

	* main/Makefile.in: -lnetwork -> @SOCK_LIBS@ in LIBS2_beos.

2026-08-23  Araki Ken <arakiken@user.sf.net>

	* main/Makefile.in: -lsocket -> -lnetwork in LIBS2_beos.

	* etc/main, etc/main.ja, man/mlterm.1, README.ja, main_loop.c:
	  Fix descriptions for font_size_range option.
	  (https://github.com/arakiken/mlterm/pull/175)

	* ReleaseNote, README.ja, man/mlterm.1: Update.

	* vt_parser.c: Fix indeterminate return value of check_ignored_csi().

	* ui_main_config.c, main_loop.c:
	  Move --igncsi / ignored_csi_list option from ui_main_config.c to
	  main_loop.c.

2026-08-22  Araki Ken <arakiken@user.sf.net>

	* ci.yml: Fix build-solaris and build-haikuos.

	* vt_term.c: Skip TEST_dynamic_comb() unless mlterm is build with bidi
	  support.

	* im_ibus.c: Call ibus_input_context_set_cursor_location() in
	  key_event() for mozc.

2026-08-16  Araki Ken <arakiken@user.sf.net>

	* vt_parser.[ch]: Add vt_set_ignored_csi_list() which sets sequence list
	  to ignore csi.

	* ui_main_config.c: Add --igncsi/ignored_csi_list option. 

2026-08-14  Araki Ken <arakiken@user.sf.net>

	* ui_im_candidate_screen.[ch]:
	  Change the type of the 2nd argument of ui_im_candidate_screen_t::select
	  and ui_im_candidate_screen_t::index from u_int to int to support
	  minus value.  (for fcitx5 + mozc)

	* ci.yml: Add build-haikuos and build-solaris.

2026-08-12  Araki Ken <arakiken@user.sf.net>

	* ui_screen.c: Revert draw_preedit_str() to the one in
	  2026/08/10 (a04071a41d38a4dc249698509263be8e82fbb47f) and
	  call redraw_screen() instead of ui_window_update() in kmsdrm alone.

2026-08-11  Araki Ken <arakiken@user.sf.net>

	* ui_screen.c: Fix unexpectedly disappearing cursor issue in inputting
	  text with the use of a input method plugin.

	* win32/ui_window.c: Fix division-by-zero error.

2026-08-10  Araki Ken <arakiken@user.sf.net>

	* ui_screen.c: ui_window_update() -> redraw_screen() in
	  draw_preedit_str() to avoid screen flickering in linux/kmsdrm.

	* ui_window.[ch]: Add ui_window_flush() for mlterm-fb in linux/kmsdrm.

	* ui_display_linux.c: ui_display_update() -> ui_display_flush(). 

2026-08-09  araki ken <arakiken@user.sf.net>

	* fb/ui_display.[ch], fb/ui.h, fb/ui_window.c:
	  #ifdef USE_KMSDRM -> #if defined(USE_KMSDRM) && defined(__linux__)

2026-08-09  Araki Ken <arakiken@user.sf.net>

	* doc/en/README.fb, doc/ja/README.fb: Update.


vtemu/mlterm.c
/* -*- c-basic-offset:2; tab-width:2; indent-tabs-mode:nil -*- */

#include <pobl/bl_config.h> /* HAVE_WINDOWS_H, WORDS_BIGENDIAN */
#ifdef HAVE_WINDOWS_H
#include <windows.h> /* In Cygwin <windows.h> is not included and error happens in jni.h. */
#endif

#include "mlterm.h"

#include <unistd.h> /* select */
#include <pobl/bl_locale.h>
#include <pobl/bl_debug.h>
#include <pobl/bl_sig_child.h>
#include <pobl/bl_mem.h> /* alloca */
#include <pobl/bl_str.h> /* bl_str_to_uint */
#include <pobl/bl_conf.h>
#include <pobl/bl_conf_io.h>
#include <pobl/bl_path.h>   /* bl_basename */
#include <pobl/bl_args.h>   /* _bl_arg_str_to_array */
#include "vt_term_manager.h"

#if defined(USE_WIN32API)
#define CONFIG_PATH "."
#elif defined(SYSCONFDIR)
#define CONFIG_PATH SYSCONFDIR
#else
#define CONFIG_PATH "/etc"
#endif

#if 0
#define USE_LOCAL_ECHO_BY_DEFAULT
#endif

#if 0
#define __DEBUG
#endif

/* --- global functions --- */

vt_term_t *mlterm_open(const char *host, const char *pass, int cols, int rows, u_int log_size,
                       const char *encoding_str, char **argv,
                       vt_xterm_event_listener_t *xterm_listener,
                       vt_config_event_listener_t *config_listener,
                       vt_screen_event_listener_t *screen_listener,
                       vt_pty_event_listener_t *pty_listener, int open_pty) {
  vt_term_t *term;
  vt_char_encoding_t encoding;
  int is_auto_encoding;
  char *envv[4];
  char *cmd_path;

  static int wsa_inited = 0;
  static u_int tab_size;
  static vt_char_encoding_t encoding_default;
  static int is_auto_encoding_default;
  static vt_unicode_policy_t unicode_policy;
  static u_int col_size_a;
  static int use_char_combining;
  static int use_multi_col_char;
  static int use_login_shell;
  static int logging_vt_seq;
  static int use_local_echo;
  static int use_locked_title;
  static int use_auto_detect;
  static int use_ansi_colors;
  static char *term_type;
  static char *public_key;
  static char *private_key;
  static char **default_argv;
  static char *default_cmd_path;

  if (!wsa_inited) {
    bl_conf_t *conf;

#if defined(USE_WIN32API) && defined(USE_LIBSSH2)
    WSADATA wsadata;
    WSAStartup(MAKEWORD(2, 0), &wsadata);

    /*
     * Prevent vt_pty_ssh from spawning a thread to watch pty.
     * If both vt_pty_ssh and MLTerm.java spawn threads to watch ptys,
     * problems will happen in reloading (that is, finalizing and
     * starting simultaneously) an applet page (mltermlet.html)
     * in a web browser.
     *
     * This hack doesn't make sense to vt_pty_pipewin32.c.
     * In the first place, if vt_pty_pipewin32 is used,
     * Java_mlterm_MLTermPty_waitForReading() returns JNI_FALSE
     * and PtyWather thread of MLTerm.java immediately exits.
     * So threads in vt_pty_pipewin32 should not be prevented.
     */
    CreateEvent(NULL, FALSE, FALSE, "PTY_READ_READY");
#endif

    bl_init_prog("mlterm", "3.9.5");
    bl_set_sys_conf_dir(CONFIG_PATH);
    bl_locale_init("");
    /*
     * :terminal with libvterm compatible library on neovim 0.5.0 stops
     * at exit if SIGCHLD handler is set in libvterm.
     */
#ifndef NO_SIGNAL
    bl_sig_child_start();
#endif
    vt_term_manager_init(1);
    vt_color_config_init();

    tab_size = 8;
    encoding_default = vt_get_char_encoding("auto");
    is_auto_encoding_default = 1;

    if (strcmp(bl_get_lang(), "ja") == 0) {
      col_size_a = 2;
    } else {
      col_size_a = 1;
    }

    use_char_combining = 1;
    use_multi_col_char = 1;
#ifdef USE_LOCAL_ECHO_BY_DEFAULT
    use_local_echo = 1;
#endif
    use_locked_title = 0;
    use_ansi_colors = 1;

    if ((conf = bl_conf_new())) {
      char *rcpath;
      char *value;

      if ((rcpath = bl_get_sys_rc_path("mlterm/main"))) {
        bl_conf_read(conf, rcpath);
        free(rcpath);
      }

      if ((rcpath = bl_get_user_rc_path("mlterm/main"))) {
        bl_conf_read(conf, rcpath);
        free(rcpath);
      }

      if ((value = bl_conf_get_value(conf, "logging_msg")) && strcmp(value, "false") == 0) {
        bl_set_msg_log_file_name(NULL);
      } else {
        bl_set_msg_log_file_name("mlterm/msg.log");
      }

      if ((value = bl_conf_get_value(conf, "logging_vt_seq")) && strcmp(value, "true") == 0) {
        logging_vt_seq = 1;
      }

      if ((value = bl_conf_get_value(conf, "tabsize"))) {
        bl_str_to_uint(&tab_size, value);
      }

      if ((value = bl_conf_get_value(conf, "encoding"))) {
        vt_char_encoding_t e;
        if ((e = vt_get_char_encoding(value)) != VT_UNKNOWN_ENCODING) {
          encoding_default = e;

          if (strcmp(value, "auto") == 0) {
            is_auto_encoding_default = 1;
          } else {
            is_auto_encoding_default = 0;
          }
        }
      }

      if ((value = bl_conf_get_value(conf, "not_use_unicode_font"))) {
        if (strcmp(value, "true") == 0) {
          unicode_policy = NOT_USE_UNICODE_FONT;
        }
      }

      if ((value = bl_conf_get_value(conf, "only_use_unicode_font"))) {
        if (strcmp(value, "true") == 0) {
          if (unicode_policy == NOT_USE_UNICODE_FONT) {
            unicode_policy = 0;
          } else {
            unicode_policy = ONLY_USE_UNICODE_FONT;
          }
        }
      }

      if ((value = bl_conf_get_value(conf, "col_size_of_width_a"))) {
        bl_str_to_uint(&col_size_a, value);
      }

      if ((value = bl_conf_get_value(conf, "use_combining"))) {
        if (strcmp(value, "false") == 0) {
          use_char_combining = 0;
        }
      }

      /* use_multi_col_char=false causes corrupt screen on SWT or vim terminal. */
#if 0
      if ((value = bl_conf_get_value(conf, "use_muti_col_char"))) {
        if (strcmp(value, "false") == 0) {
          use_multi_col_char = 0;
        }
      }
#endif

      if ((value = bl_conf_get_value(conf, "use_login_shell"))) {
        if (strcmp(value, "true") == 0) {
          use_login_shell = 1;
        }
      }

      if ((value = bl_conf_get_value(conf, "use_local_echo"))) {
#ifdef USE_LOCAL_ECHO_BY_DEFAULT
        if (strcmp(value, "false") == 0) {
          use_local_echo = 0;
        }
#else
        if (strcmp(value, "true") == 0) {
          use_local_echo = 1;
        }
#endif
      }

      if ((value = bl_conf_get_value(conf, "use_locked_title"))) {
        if (strcmp(value, "true") == 0) {
            use_locked_title = 1;
        }
      }

      if ((value = bl_conf_get_value(conf, "use_alt_buffer"))) {
        if (strcmp(value, "false") == 0) {
          vt_set_use_alt_buffer(0);
        }
      }

      if ((value = bl_conf_get_value(conf, "use_ansi_colors"))) {
        if (strcmp(value, "false") == 0) {
          use_ansi_colors = 0;
        }
      }

      if ((value = bl_conf_get_value(conf, "auto_detect_encodings"))) {
        vt_set_auto_detect_encodings(value);
      }

      if ((value = bl_conf_get_value(conf, "use_auto_detect"))) {
        if (strcmp(value, "true") == 0) {
          use_auto_detect = 1;
        }
      }

      if ((value = bl_conf_get_value(conf, "termtype"))) {
        term_type = strdup(value);
      } else {
        term_type = "xterm";
      }

#ifdef USE_LIBSSH2
      if ((value = bl_conf_get_value(conf, "ssh_public_key"))) {
        public_key = strdup(value);
      }

      if ((value = bl_conf_get_value(conf, "ssh_private_key"))) {
        private_key = strdup(value);
      }

      if ((value = bl_conf_get_value(conf, "cipher_list"))) {
        vt_pty_ssh_set_cipher_list(strdup(value));
      }

      if ((value = bl_conf_get_value(conf, "ssh_keepalive_interval"))) {
        u_int keepalive_interval;

        if (bl_str_to_uint(&keepalive_interval, value) && keepalive_interval > 0) {
          vt_pty_ssh_set_keepalive_interval(keepalive_interval);
        }
      }

      if ((value = bl_conf_get_value(conf, "ssh_x11_forwarding"))) {
        if (strcmp(value, "true") == 0) {
          vt_pty_ssh_set_use_x11_forwarding(NULL, 1);
        }
      }

      if ((value = bl_conf_get_value(conf, "allow_scp"))) {
        if (strcmp(value, "true") == 0) {
          vt_set_use_scp_full(1);
        }
      }
#endif

#if 0
      /* XXX How to get password ? */
      if ((value = bl_conf_get_value(conf, "default_server"))) {
        default_server = strdup(value);
      }
#endif

      if (((value = bl_conf_get_value(conf, "exec_cmd"))
#if defined(USE_WIN32API) && !defined(USE_LIBSSH2)
           || (value = plink)
#endif
               ) &&
          (default_argv = malloc(sizeof(char *) * bl_count_char_in_str(value, ' ') + 2))) {
        int argc;

        bl_arg_str_to_array(default_argv, &argc, strdup(value));
        default_cmd_path = default_argv[0];
      }

      bl_conf_destroy(conf);
    }

    unicode_policy |= ONLY_USE_UNICODE_BOXDRAW_FONT;

    wsa_inited = 1;
  }

  encoding = encoding_default;
  is_auto_encoding = is_auto_encoding_default;

  if (encoding_str) {
    vt_char_encoding_t e;

    if ((e = vt_get_char_encoding(encoding_str)) != VT_UNKNOWN_ENCODING) {
      encoding = e;

      if (strcmp(encoding_str, "auto") == 0) {
        is_auto_encoding = 1;
      } else {
        is_auto_encoding = 0;
      }
    }
  }

  if (!(term = vt_create_term(term_type, cols, rows, tab_size, log_size, encoding, is_auto_encoding,
                              use_auto_detect, logging_vt_seq, unicode_policy, col_size_a,
                              use_char_combining, use_multi_col_char, 0 /* use_ctl */,
                              0 /* bidi_mode */, NULL /* bidi_separators */,
                              0 /* use_dynamic_comb */, BSM_STATIC, 0 /* vertical_mode */,
                              use_local_echo, NULL, NULL, use_ansi_colors, 0 /* alt_color_mode */,
                              0 /* use_ot_layout */, 0 /* cursor_style */,
                              0 /* ignore_broadcasted_chars */,
                              use_locked_title))) {
    goto error;
  }

  vt_term_attach(term, xterm_listener, config_listener, screen_listener, pty_listener);

  if (!open_pty) {
    return term;
  }

  if (!host && !(host = getenv("DISPLAY"))) {
    host = ":0.0";
  }

  if (argv) {
    cmd_path = argv[0];
  } else if (default_argv) {
    argv = default_argv;
    cmd_path = default_cmd_path;
  } else {
#ifndef USE_WIN32API
    if (pass)
#endif
    {
      cmd_path = NULL;
      argv = alloca(sizeof(char *));
      argv[0] = NULL;
    }
#ifndef USE_WIN32API
    else {
      cmd_path = getenv("SHELL");
      argv = alloca(sizeof(char *) * 2);
      argv[1] = NULL;
    }
#endif
  }

  if (cmd_path) {
    if (use_login_shell) {
      argv[0] = alloca(strlen(cmd_path) + 2);
      sprintf(argv[0], "-%s", bl_basename(cmd_path));
    } else {
      argv[0] = bl_basename(cmd_path);
    }
  }

  envv[0] = alloca(8 + strlen(host) + 1);
  sprintf(envv[0], "DISPLAY=%s", host);
  envv[1] = "TERM=xterm";
  envv[2] = "COLORFGBG=default;default";
  envv[3] = NULL;

  if (vt_term_open_pty(term, cmd_path, argv, envv, host, NULL, pass, public_key, private_key,
                       0, 0)) {
    return term;
  } else {
    vt_destroy_term(term);
  }

error:
#ifdef __DEBUG
  bl_debug_printf(BL_DEBUG_TAG " Failed to open pty.\n");
#endif

  return NULL;
}

void mlterm_final(void) {
  vt_term_manager_final();
  vt_color_config_final();
  vt_free_word_separators();
#ifndef NO_SIGNAL
  bl_sig_child_final();
#endif
  bl_locale_final();
}
